### PR TITLE
Feature/thread safe

### PR DESCRIPTION
**Pull request by SalvoVirga**
simple_msgs are sockets are now thread safe

### CODE REVIEW

File: include/simple/generic_socket.hpp [hunk 1]
```diff
@@ -50,16 +51,20 @@ class GenericSocket {
    * @brief Move constructor.
    */
   GenericSocket(GenericSocket&& other) {
-    socket_.store(other.socket_);
+    std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline include/simple/generic_socket.hpp]: I think you should use std::lock here to be fully threadsafe. The chance is pretty slim but there is still potential for a deadlock here if another thread does the same thing on the other object..
```
        std::lock(mutex_, other.mutex_);
        std::lock_guard<std::mutex> lock(mutex_, std::adopt_lock);
        std::lock_guard<std::mutex> other_lock(other.mutex_, std::adopt_lock);
```

File: include/simple/generic_socket.hpp [hunk 2]
```diff
@@ -50,16 +51,20 @@ class GenericSocket {
...
+    socket_ = other.socket_;
     other.socket_ = nullptr;
   }
 
   /**
    * @brief Move assignment operator.
    */
   GenericSocket& operator=(GenericSocket&& other) {
+    std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline include/simple/generic_socket.hpp]: same as above, std::lock should make it even more safe

File: include/simple/generic_socket.hpp [hunk 3]
```diff
@@ -298,8 +321,9 @@ class GenericSocket {
   inline const std::string& endpoint() { return endpoint_; }
 
 private:
-  std::string topic_{T::getTopic()};    //! The message topic, internally defined for each SIMPLE message.
-  std::atomic<void*> socket_{nullptr};  //! The internal ZMQ socket.
+  mutable std::mutex mutex_{};        //! Mutex for thread-safety.
+  std::string topic_{T::getTopic()};  //! The message topic, internally defined for each SIMPLE message.
```
Comment (TheHugeManatee) [inline include/simple/generic_socket.hpp]: could this be const?

File: include/simple/server.hpp [hunk 4]
```diff
@@ -132,11 +133,11 @@ class Server {
    * callback function and reply.
    */
   void awaitRequest(std::shared_ptr<std::atomic<bool>> alive, std::shared_ptr<GenericSocket<T>> socket) {
-    while (*alive) {
+    while (alive->load()) {
```
Comment (TheHugeManatee) [inline include/simple/server.hpp:136]: The previous version was as good, but I guess you were going for a consistent use of load()/store() for all atomics?

File: include/simple/server.hpp [hunk 5]
```diff
@@ -132,11 +133,11 @@ class Server {
    * callback function and reply.
    */
   void awaitRequest(std::shared_ptr<std::atomic<bool>> alive, std::shared_ptr<GenericSocket<T>> socket) {
```
Comment (TheHugeManatee) [inline include/simple/server.hpp:135]: Not directly related to this PR, but still: Why exactly are the two member functions passed again to the thread function as parameters? They should be accessible anyways directly, and passing them only means increasing the refcount. Also, why are they shared_ptr's in the first place? I can see no good reason for it, they should live as long as the thread itself anyways?

Comment (TheHugeManatee) [reply include/simple/server.hpp:135]: Ah, might be that this was an issue before, when you did `detach()` instead of `join()` in the destructor. Now it should be guaranteed that the alive_ and socket_ live longer than the thread, so direct access should be possible?

File: msgs/include/simple_msgs/image.hpp [hunk 6]
```diff
@@ -77,6 +77,7 @@ class Image : public GenericMessage {
   Image& operator=(const Image& other) {
     if (this != std::addressof(other)) {
       std::lock_guard<std::mutex> lock{mutex_};
+      std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/image.hpp]: Same issue as before, use std::lock here

File: msgs/include/simple_msgs/image.hpp [hunk 7]
```diff
@@ -99,6 +100,7 @@ class Image : public GenericMessage {
   Image& operator=(Image&& other) noexcept {
     if (this != std::addressof(other)) {
       std::lock_guard<std::mutex> lock{mutex_};
+      std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/image.hpp]: Same issue as before, use std::lock here

File: include/simple/server.hpp [hunk 8]
```diff
@@ -53,7 +53,8 @@ class Server {
    * @param [in] linger - Time the unsent messages linger in memory after the socket
```
Comment (TheHugeManatee) [inline include/simple/server.hpp:53]: There's some copy-past mistakes in the comments of this file (which I can't mark directly because they are not changed) which mention "Subscriber" instead of "Server". I saw some in the move assignment operator and in the comment for `isValid()`, there may be more

File: msgs/include/simple_msgs/numeric_type.hpp [hunk 9]
```diff
@@ -140,7 +139,7 @@ class NumericType : public GenericMessage {
    * @brief Addition operator.
    */
   NumericType& operator+=(const NumericType& rhs) {
-    data_ += rhs.data_;
+    data_.exchange(data_.load() + rhs.data_.load());
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/numeric_type.hpp]: Don't want to shit on work that's done already but this would be a good point to ditch the math functionality from your library. Ultimately I think it's not necessary to have these operator overloads (everyone will have their own math types anyways) and it would reduce your codebase, making it easier to maintain. It will not really be a big deal either way but to me it feels cleaner not to include math in a networking lib... Just my two cents. Also, this operation here is not strictly fully threadsafe, and there's no way to fix it unless you put on a full mutex here... :-/

File: msgs/include/simple_msgs/pose_stamped.h [hunk 10]
```diff
@@ -70,7 +71,10 @@ class PoseStamped : public GenericMessage {
   /**
    * @brief Returns true if lhs is equal to rhs, false otherwise.
    */
-  inline bool operator==(const PoseStamped& rhs) const { return (pose_ == rhs.pose_ && header_ == rhs.header_); }
+  inline bool operator==(const PoseStamped& rhs) const {
+    std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/pose_stamped.h:75]: should also lock the rhs here, also using `std::lock` as above

File: msgs/include/simple_msgs/quaternion.h [hunk 11]
```diff
@@ -87,7 +88,10 @@ class Quaternion : public GenericMessage {
   /**
    * @brief Returns true if lhs is equal to rhs, false otherwise.
    */
-  inline bool operator==(const Quaternion& rhs) const { return data_ == rhs.data_; }
+  inline bool operator==(const Quaternion& rhs) const {
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/quaternion.h:91]: also lock `rhs` here, with `std::lock`

File: msgs/include/simple_msgs/quaternion_stamped.h [hunk 12]
```diff
@@ -73,6 +73,7 @@ class QuaternionStamped : public GenericMessage {
    * @brief Returns true if lhs is equal to rhs, false otherwise.
    */
   inline bool operator==(const QuaternionStamped& rhs) const {
+    std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/quaternion_stamped.h]: lock `rhs` here as well

File: msgs/include/simple_msgs/rotation_matrix.h [hunk 13]
```diff
@@ -90,7 +91,10 @@ class RotationMatrix : public GenericMessage {
   /**
    * @brief Returns true if lhs is equal to rhs, false otherwise.
    */
-  inline bool operator==(const RotationMatrix& rhs) const { return data_ == rhs.data_; }
+  inline bool operator==(const RotationMatrix& rhs) const {
+    std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/rotation_matrix.h]: lock `rhs` here

File: msgs/include/simple_msgs/rotation_matrix_stamped.h [hunk 14]
```diff
@@ -73,6 +73,7 @@ class RotationMatrixStamped : public GenericMessage {
    * @brief Returns true if lhs is equal to rhs, false otherwise.
    */
   inline bool operator==(const RotationMatrixStamped& rhs) const {
+    std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/include/simple_msgs/rotation_matrix_stamped.h]: lock `rhs.mutex_`

File: msgs/src/header.cpp [hunk 15]
```diff
@@ -22,11 +22,12 @@ Header::Header(const void* data)
 Header::Header(const Header& h) : Header{h.seq_n_, h.frame_id_, h.timestamp_} {}
 
 Header::Header(Header&& other) noexcept
-  : seq_n_{std::move(other.seq_n_)}, frame_id_(std::move(other.frame_id_)), timestamp_{std::move(other.timestamp_)} {}
+  : seq_n_{other.seq_n_}, frame_id_(std::move(other.frame_id_)), timestamp_{other.timestamp_} {}
```
Comment (TheHugeManatee) [inline msgs/src/header.cpp]: i think there should be some locking here..

File: msgs/src/header.cpp [hunk 16]
```diff
@@ -22,11 +22,12 @@ Header::Header(const void* data)
...
 
 Header& Header::operator=(const Header& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/header.cpp]: use `std::lock`

File: msgs/src/header.cpp [hunk 17]
```diff
@@ -36,9 +37,11 @@ Header& Header::operator=(const Header& other) {
 
 Header& Header::operator=(Header&& other) noexcept {
   if (this != std::addressof(other)) {
-    seq_n_ = std::move(other.seq_n_);
+    std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/header.cpp]: use `std::lock`

File: msgs/src/header.cpp [hunk 18]
```diff
@@ -63,10 +66,7 @@ std::shared_ptr<flatbuffers::DetachedBuffer> Header::getBufferData() const {
   return std::make_shared<flatbuffers::DetachedBuffer>(builder.Release());
 }
 
-void Header::setSequenceNumber(int seq_n) {
-  std::lock_guard<std::mutex> lock{mutex_};
-  seq_n_ = seq_n;
-}
+void Header::setSequenceNumber(int seq_n) { seq_n_ = seq_n; }
```
Comment (TheHugeManatee) [inline msgs/src/header.cpp]: why did you remove the lock here?

File: msgs/src/point.cpp [hunk 19]
```diff
@@ -33,6 +33,7 @@ Point::Point(Point&& other) noexcept : data_{std::move(other.data_)} {}
 Point& Point::operator=(const Point& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/point.cpp]: use std::lock

File: msgs/src/point.cpp [hunk 20]
```diff
@@ -41,6 +42,7 @@ Point& Point::operator=(const Point& other) {
 Point& Point::operator=(Point&& other) noexcept {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/point.cpp]: use std::lock

File: msgs/src/point_stamped.cpp [hunk 21]
```diff
@@ -26,6 +26,7 @@ PointStamped::PointStamped(PointStamped&& other) noexcept
 PointStamped& PointStamped::operator=(const PointStamped& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/point_stamped.cpp]: use std;:lock

File: msgs/src/point_stamped.cpp [hunk 22]
```diff
@@ -35,6 +36,7 @@ PointStamped& PointStamped::operator=(const PointStamped& other) {
 PointStamped& PointStamped::operator=(PointStamped&& other) noexcept {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/point_stamped.cpp]: use std::lock

File: msgs/src/pose.cpp [hunk 23]
```diff
@@ -23,20 +23,22 @@ Pose::Pose(const Pose& other) : Pose{other.position_, other.quaternion_} {}
...
 Pose::Pose(Pose&& other) noexcept : Pose{std::move(other.position_), std::move(other.quaternion_)} {}
 
-Pose& Pose::operator=(const Pose& p) {
-  if (this != std::addressof(p)) {
+Pose& Pose::operator=(const Pose& other) {
+  if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
-    position_ = p.position_;
-    quaternion_ = p.quaternion_;
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/pose.cpp]: use std::lock

File: msgs/src/pose.cpp [hunk 24]
```diff
@@ -23,20 +23,22 @@ Pose::Pose(const Pose& other) : Pose{other.position_, other.quaternion_} {}
...
 }
 
-Pose& Pose::operator=(Pose&& p) noexcept {
-  if (this != std::addressof(p)) {
+Pose& Pose::operator=(Pose&& other) noexcept {
+  if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
-    position_ = std::move(p.position_);
-    quaternion_ = std::move(p.quaternion_);
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/pose.cpp]: use std::lock

File: msgs/src/pose_stamped.cpp [hunk 25]
```diff
@@ -24,20 +24,22 @@ PoseStamped::PoseStamped(const PoseStamped& other) : PoseStamped{other.header_,
...
   : PoseStamped{std::move(other.header_), std::move(other.pose_)} {}
 
-PoseStamped& PoseStamped::operator=(const PoseStamped& p) {
-  if (this != std::addressof(p)) {
+PoseStamped& PoseStamped::operator=(const PoseStamped& other) {
+  if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
-    header_ = p.header_;
-    pose_ = p.pose_;
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/pose_stamped.cpp]: use std::lock

File: msgs/src/pose_stamped.cpp [hunk 26]
```diff
@@ -24,20 +24,22 @@ PoseStamped::PoseStamped(const PoseStamped& other) : PoseStamped{other.header_,
...
 }
 
-PoseStamped& PoseStamped::operator=(PoseStamped&& p) noexcept {
-  if (this != std::addressof(p)) {
+PoseStamped& PoseStamped::operator=(PoseStamped&& other) noexcept {
+  if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
-    header_ = std::move(p.header_);
-    pose_ = std::move(p.pose_);
+    std::lock_guard<std::mutex> other_lock{other.mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/pose_stamped.cpp]: use std::lock

File: msgs/src/quaternion.cpp [hunk 27]
```diff
@@ -32,6 +32,7 @@ Quaternion::Quaternion(Quaternion&& other) noexcept : data_{std::move(other.data
 Quaternion& Quaternion::operator=(const Quaternion& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/quaternion.cpp]: use std::lock

File: msgs/src/quaternion.cpp [hunk 28]
```diff
@@ -40,6 +41,7 @@ Quaternion& Quaternion::operator=(const Quaternion& other) {
 Quaternion& Quaternion::operator=(Quaternion&& other) noexcept {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/quaternion.cpp]: use std::lock

File: msgs/src/quaternion_stamped.cpp [hunk 29]
```diff
@@ -30,6 +30,7 @@ QuaternionStamped::QuaternionStamped(QuaternionStamped&& other) noexcept
 QuaternionStamped& QuaternionStamped::operator=(const QuaternionStamped& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/quaternion_stamped.cpp]: use std::lock

File: msgs/src/quaternion_stamped.cpp [hunk 30]
```diff
@@ -39,6 +40,7 @@ QuaternionStamped& QuaternionStamped::operator=(const QuaternionStamped& other)
 QuaternionStamped& QuaternionStamped::operator=(QuaternionStamped&& other) noexcept {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/quaternion_stamped.cpp]: use std::lock

File: msgs/src/rotation_matrix.cpp [hunk 31]
```diff
@@ -41,6 +41,7 @@ RotationMatrix::RotationMatrix(RotationMatrix&& m) noexcept : RotationMatrix{std
 RotationMatrix& RotationMatrix::operator=(const RotationMatrix& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/rotation_matrix.cpp]: use std::lock

File: msgs/src/rotation_matrix.cpp [hunk 32]
```diff
@@ -49,6 +50,7 @@ RotationMatrix& RotationMatrix::operator=(const RotationMatrix& other) {
 RotationMatrix& RotationMatrix::operator=(RotationMatrix&& other) noexcept {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/rotation_matrix.cpp]: use std::lock

File: msgs/src/rotation_matrix_stamped.cpp [hunk 33]
```diff
@@ -30,6 +30,7 @@ RotationMatrixStamped::RotationMatrixStamped(RotationMatrixStamped&& other) noex
 RotationMatrixStamped& RotationMatrixStamped::operator=(const RotationMatrixStamped& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/rotation_matrix_stamped.cpp]: use std::lock

File: msgs/src/rotation_matrix_stamped.cpp [hunk 34]
```diff
@@ -39,6 +40,7 @@ RotationMatrixStamped& RotationMatrixStamped::operator=(const RotationMatrixStam
 RotationMatrixStamped& RotationMatrixStamped::operator=(RotationMatrixStamped&& other) noexcept {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/rotation_matrix_stamped.cpp]: use std::lock

File: msgs/src/string.cpp [hunk 35]
```diff
@@ -26,6 +26,7 @@ String::String(String&& other) noexcept : data_{std::move(other.data_)} {}
 String& String::operator=(const String& other) {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/string.cpp]: use std::lock

File: msgs/src/string.cpp [hunk 36]
```diff
@@ -34,6 +35,7 @@ String& String::operator=(const String& other) {
 String& String::operator=(String&& other) noexcept {
   if (this != std::addressof(other)) {
     std::lock_guard<std::mutex> lock{mutex_};
```
Comment (TheHugeManatee) [inline msgs/src/string.cpp]: use std::lock

**Review by TheHugeManatee — CHANGES_REQUESTED** [review]
Looks mostly good, but I think you should use std::lock for all occasions where you lock two locks at once, to guarantee no deadlocks. I marked all occasions I saw with a comment..

File: include/simple/server.hpp (hunk 4 above)
Comment (SalvoVirga) [reply include/simple/server.hpp:136]: yes, just a matter of consistency

File: include/simple/generic_socket.hpp (hunk 3 above)
Comment (SalvoVirga) [reply include/simple/generic_socket.hpp]: made it const

File: msgs/include/simple_msgs/pose_stamped.h (hunk 10 above)
Comment (SalvoVirga) [reply msgs/include/simple_msgs/pose_stamped.h:75]: 👍

File: msgs/include/simple_msgs/quaternion_stamped.h (hunk 12 above)
Comment (SalvoVirga) [reply msgs/include/simple_msgs/quaternion_stamped.h]: 👍

File: msgs/include/simple_msgs/quaternion.h (hunk 11 above)
Comment (SalvoVirga) [reply msgs/include/simple_msgs/quaternion.h:91]: 👍

File: msgs/include/simple_msgs/rotation_matrix.h (hunk 13 above)
Comment (SalvoVirga) [reply msgs/include/simple_msgs/rotation_matrix.h]: 👍

File: msgs/src/header.cpp (hunk 15 above)
Comment (SalvoVirga) [reply msgs/src/header.cpp]: This made me realize that some locking is needed in all the copy/move constructors of simple_msgs...

I'll add private constructors that take the object to be copied/move and a lock_guard, public copy/move constructors will call those passing over the object (other) and a lock_guard that uses the other.mutex_.

File: msgs/include/simple_msgs/numeric_type.hpp (hunk 9 above)
Comment (SalvoVirga) [reply msgs/include/simple_msgs/numeric_type.hpp]: I looked up around to find a nicer way to fix this, but now I have to agree with you :\ I'll remove the arithmetical operations and that's it

File: include/simple/server.hpp (hunk 5 above)
Comment (SalvoVirga) [reply include/simple/server.hpp:135]: Correct! Now that the two threads are joined, we can simplify that function again. The same for the subscriber by the way. Thanks for noticing it

**Review by TheHugeManatee — APPROVED** [review]
The idiom for locked moving is pretty nice!
